server/router/api/v1/thumbnail.go

```diff
@@ -0,0 +1,146 @@
+package v1
+
+import (
+	"bytes"
+	"fmt"
+	"image"
+	"io"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	"github.com/disintegration/imaging"
+	"github.com/pkg/errors"
+
+	"github.com/usememos/memos/store"
+)
+
+// thumbnail provides functionality to manage thumbnail images
+// for resources.
+type thumbnail struct {
+	// The resource the thumbnail is for
+	resource *store.Resource
+}
+
+func (thumbnail) supportedMimeTypes() []string {
+	return []string{
+		"image/png",
+		"image/jpeg",
+	}
+}
+
+func (t *thumbnail) getFilePath(assetsFolderPath string) (string, error) {
+	if assetsFolderPath == "" {
+		return "", errors.New("aapplication path is not set")
+	}
+
+	ext := filepath.Ext(t.resource.Filename)
+	path := filepath.Join(assetsFolderPath, thumbnailImagePath, fmt.Sprintf("%d%s", t.resource.ID, ext))
+
+	return path, nil
+}
+
+func (t *thumbnail) getFile(assetsFolderPath string) ([]byte, error) {
+	path, err := t.getFilePath(assetsFolderPath)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get thumbnail file path")
+	}
+
+	if _, err := os.Stat(path); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, errors.Wrap(err, "failed to check thumbnail image stat")
+		}
+	}
+
+	dstFile, err := os.Open(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open thumbnail file")
+	}
+	defer dstFile.Close()
+
+	dstBlob, err := io.ReadAll(dstFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read thumbnail file")
+	}
+
+	return dstBlob, nil
+}
+
+func (thumbnail) generateImage(sourceBlob []byte) (image.Image, error) {
+	var availableGeneratorAmount int32 = 32
+
+	if atomic.LoadInt32(&availableGeneratorAmount) <= 0 {
+		return nil, errors.New("not enough available generator amount")
+	}
+
+	atomic.AddInt32(&availableGeneratorAmount, -1)
+	defer func() {
+		atomic.AddInt32(&availableGeneratorAmount, 1)
+	}()
+
+	reader := bytes.NewReader(sourceBlob)
+	src, err := imaging.Decode(reader, imaging.AutoOrientation(true))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode thumbnail image")
+	}
+
+	thumbnailImage := imaging.Resize(src, 512, 0, imaging.Lanczos)
```

Comment:
Thank you for working on this, but I have a couple of concerns:
1. Images in the web client have a maximum width of `700px` at 100% zoom. When the thumbnail width is limited to `512px`, the images are no longer full width. Is there a reason behind this number?
2. Images less than 512px wide are upscaled. Maybe omit thumbnail generation in that case?

Reply:
Good catch

the thumbnail generation code was a lift and shift from v0.21, this is where the 512px comes from. I had a look through the history and when the code was added in v0.13 the timeline images had a max width of 320px so it worked in that scenario. Over the following versions timeline images got larger and the thumbnail flag was remove from them and only the resource page images, where images are < 512px, kept the flag

I'll update the code based on your feedback